Mathlib/LinearAlgebra/PerfectPairing/Basic.lean

```diff
@@ -82,6 +84,24 @@ protected instance IsPerfPair.id [IsReflexive R M] : IsPerfPair (.id (R := R) (M
 /-- A reflexive module has a perfect pairing with its dual. -/
 instance IsPerfPair.dualEval [IsReflexive R M] : IsPerfPair (Dual.eval R M) := .flip .id
 
+instance IsPerfPair.compl₁₂ (eM : M' ≃ₗ[R] M) (eN : N' ≃ₗ[R] N) :
+    (p.compl₁₂ eM eN : M' →ₗ[R] N' →ₗ[R] R).IsPerfPair :=
+  ⟨((LinearEquiv.congrLeft R R eN).symm.bijective.comp
+    (IsPerfPair.bijective_left p)).comp eM.bijective,
+    ((LinearEquiv.congrLeft R R eM).symm.bijective.comp
+    (IsPerfPair.bijective_right p)).comp eN.bijective⟩
+
+lemma IsPerfPair.congr (eM : M' ≃ₗ[R] M) (eN : N' ≃ₗ[R] N) (q : M' →ₗ[R] N' →ₗ[R] R)
+    (H : q.compl₁₂ eM.symm eN.symm = p) : q.IsPerfPair := by
+  obtain rfl : q = p.compl₁₂ eM eN := by subst H; ext; simp
+  infer_instance
```

Comment:
Do we definitely want this lemma?

A hypothesis like `H` containing an equality says to me that the user should just have rewritten first?

Reply:
Yes, but no, because the point is to have the hypothesis `p.IsPerfPair` inferred by TC search. The deeper reason why the usual rules don't apply is that now `IsPerfPair` is a typeclass on terms, a "termclass" so to speak, and that's quite rare.